suite/tests/client-interface/blackbox.dll.c

```diff
@@ -0,0 +1,72 @@
+ * Copyright (c) 2019 Google, Inc. All rights reserved.
+ * *******************************************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of MIT nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL MIT OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* Test exposed instr_t size to be identical to internal size by initializing an stack
+ * object with DynamoRIO and checking its bounds. TODO: Add test for opnd_t.
+ */
+
+#include "dr_api.h"
+#include "client_tools.h"
+#include <string.h>
+
+#define INSTR_LSB *(byte *)(&buf)
+#define INSTR_MSB *((byte *)(&buf) + sizeof(instr_t) - 1)
+#define REDZONE *((byte *)&buf + sizeof(instr_t))
+
+#define GENERATE_TEST(type)                \
+    char buf[sizeof(type) + 1];            \
+    REDZONE = 0xcd;                        \
+    type *ins = (type *)buf;               \
+    memset(ins, 0xab, sizeof(type));       \
+    if (INSTR_LSB != 0xab) {               \
+        dr_printf("Error\n");              \
+    }                                      \
+    if (INSTR_MSB != 0xab) {               \
+        dr_printf("Error\n");              \
+    }                                      \
+    if (REDZONE != 0xcd) {                 \
+        dr_printf("Error\n");              \
+    }                                      \
+    void *dc = dr_get_current_drcontext(); \
+    instr_init(dc, ins);                   \
+    if (INSTR_MSB != 0x0) {                \
+        dr_printf("Error\n");              \
+    }                                      \
+    if (REDZONE != 0xcd) {                 \
```

Comment:
This only tests against instr_t being larger than the black box if instr_init goes and writes to the end which may not be the case.

Reply:
The != 0 check above checks for instr_t being smaller than the blackbox. I am aware this test here is perfect, but it's due to the types being abstracted in the first place. By the way, did you see, there was a bug (blackbox too big).